File: core/business-tier/src/main/scala/ru/korus/tmis/laboratory/across/request/DiagnosticRequestInfo.scala
package ru.korus.tmis.laboratory.across.request

import java.util.Date

import ru.korus.tmis.laboratory.across.{ws => lab2}


import DataConverter._
import ru.korus.tmis.core.exception.CoreException

import ru.korus.tmis.scala.util.Defaultible
import Defaultible.setDefault
import ru.korus.tmis.laboratory.across.request.Utility._

sealed case class DiagnosticRequestInfo(
                                         orderMisId: Int,
                                         orderCaseId: Option[String],
                                         orderFinanceId: Option[Int],
                                         orderMisDate: Option[Date],
                                         orderPregnatMin: Option[Int],
                                         orderPregnatMax: Option[Int],
                                         orderDiagCode: Option[String],
                                         orderDiagText: Option[String],
                                         orderComment: Option[String],
                                         orderDepartmentName: Option[String],
                                         orderDepartmentMisCode: Option[String],
                                         orderDoctorFamily: Option[String],
                                         orderDoctorName: Option[String],
                                         orderDoctorPatronum: Option[String],
                                         orderDoctorMisId: Option[Int]
                                         )

object DiagnosticRequestInfo {

  implicit def toLab2DRI(v: DiagnosticRequestInfo): lab2.DiagnosticRequestInfo = {
    implicit def locallyDefaultString = setDefault("")
    implicit def locallyDefaultInt = setDefault(-1)

    import v._
    val ret = new lab2.DiagnosticRequestInfo

    ret.setOrderMisId(orderMisId)

    setAsOptional(orderCaseId) {
      ret.setOrderCaseId(_)
    }

    setAsOptional(orderFinanceId) {
      ret.setOrderFinanceId(_)
    }

    setAsRequired(new CoreException("OrderMisDate: not found"))(orderMisDate) {
      it => ret.setOrderMisDate(date2GC(it))
    }

    for (min <- orderPregnatMin; max <- orderPregnatMax) yield ret.setOrderPregnat(min / 2 + max / 2)

    setAsOptional(orderDiagCode) {
      ret.setOrderDiagCode(_)
    }
    setAsOptional(orderDiagText) {
      ret.setOrderDiagText(_)
    }
    setAsOptional(orderComment) {
      ret.setOrderComment(_)
    }
    setAsOptional(orderDepartmentName) {
      ret.setOrderDepartmentName(_)
    }
    setAsDefaultible(orderDepartmentMisCode) {
      it => ret.setOrderDepartmentMisId(it)
    }
    setAsOptional(orderDoctorFamily) {
      ret.setOrderDoctorFamily(_)
    }
    setAsOptional(orderDoctorName) {
      ret.setOrderDoctorName(_)
    }
    setAsOptional(orderDoctorPatronum) {
      ret.setOrderDoctorPatronum(_)
    }
    setAsDefaultible(orderDoctorMisId) {
      it => ret.setOrderDoctorMisId(it.toString)
    }
    ret
  }

}

File: core/common/src/main/java/ru/korus/tmis/lis/data/DiagnosticRequestInfo.java
package ru.korus.tmis.lis.data;

import org.joda.time.DateTime;
import ru.korus.tmis.core.entity.model.Staff;

import java.io.Serializable;

/**
 * Author:      Dmitriy E. Nosov <br>
 * Date:        20.11.13, 19:38 <br>
 * Company:     Korus Consulting IT<br>
 * Description:  <br>
 */
public class DiagnosticRequestInfo implements Serializable {

    public static final long serialVersionUID = 1L;

    private int orderMisId;
    private String orderCaseId;
    private int orderFinanceId;
    private DateTime orderMisDate;
    private int orderPregnatMin;
    private int orderPregnatMax;
    private String orderDiagCode;
    private String orderDiagText;
    private String orderComment;
    private String orderDepartmentName;
    private String orderDepartmentMisCode;
    private String orderDoctorFamily;
    private String orderDoctorName;
    private String orderDoctorPatronum;
    private int orderDoctorMisId;
    private Staff orderDoctorMis;

    public DiagnosticRequestInfo() {
    }

    public int getOrderMisId() {
        return orderMisId;
    }

    public void setOrderMisId(int orderMisId) {
        this.orderMisId = orderMisId;
    }

    public String getOrderCaseId() {
        return orderCaseId;
    }

    public void setOrderCaseId(String orderCaseId) {
        this.orderCaseId = orderCaseId;
    }

    public int getOrderFinanceId() {
        return orderFinanceId;
    }

    public void setOrderFinanceId(int orderFinanceId) {
        this.orderFinanceId = orderFinanceId;
    }

    public DateTime getOrderMisDate() {
        return orderMisDate;
    }

    public void setOrderMisDate(DateTime orderMisDate) {
        this.orderMisDate = orderMisDate;
    }

    public int getOrderPregnatMin() {
        return orderPregnatMin;
    }

    public void setOrderPregnatMin(int orderPregnatMin) {
        this.orderPregnatMin = orderPregnatMin;
    }

    public int getOrderPregnatMax() {
        return orderPregnatMax;
    }

    public void setOrderPregnatMax(int orderPregnatMax) {
        this.orderPregnatMax = orderPregnatMax;
    }

    public String getOrderDiagCode() {
        return orderDiagCode;
    }

    public void setOrderDiagCode(String orderDiagCode) {
        this.orderDiagCode = orderDiagCode;
    }

    public String getOrderDiagText() {
        return orderDiagText;
    }

    public void setOrderDiagText(String orderDiagText) {
        this.orderDiagText = orderDiagText;
    }

    public String getOrderComment() {
        return orderComment;
    }

    public void setOrderComment(String orderComment) {
        this.orderComment = orderComment;
    }

    public String getOrderDepartmentName() {
        return orderDepartmentName;
    }

    public void setOrderDepartmentName(String orderDepartmentName) {
        this.orderDepartmentName = orderDepartmentName;
    }

    public String getOrderDepartmentMisCode() {
        return orderDepartmentMisCode;
    }

    public void setOrderDepartmentMisCode(String orderDepartmentMisCode) {
        this.orderDepartmentMisCode = orderDepartmentMisCode;
    }

    public String getOrderDoctorFamily() {
        return orderDoctorFamily;
    }

    public void setOrderDoctorFamily(String orderDoctorFamily) {
        this.orderDoctorFamily = orderDoctorFamily;
    }

    public String getOrderDoctorName() {
        return orderDoctorName;
    }

    public void setOrderDoctorName(String orderDoctorName) {
        this.orderDoctorName = orderDoctorName;
    }

    public String getOrderDoctorPatronum() {
        return orderDoctorPatronum;
    }

    public void setOrderDoctorPatronum(String orderDoctorPatronum) {
        this.orderDoctorPatronum = orderDoctorPatronum;
    }

    public int getOrderDoctorMisId() {
        return orderDoctorMisId;
    }

    public void setOrderDoctorMisId(int orderDoctorMisId) {
        this.orderDoctorMisId = orderDoctorMisId;
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("DiagnosticRequestInfo{");
        sb.append("orderMisId=").append(orderMisId);
        sb.append(", orderCaseId='").append(orderCaseId).append('\'');
        sb.append(", orderFinanceId=").append(orderFinanceId);
        sb.append(", orderMisDate=").append(orderMisDate);
        sb.append(", orderPregnatMin=").append(orderPregnatMin);
        sb.append(", orderPregnatMax=").append(orderPregnatMax);
        sb.append(", orderDiagCode='").append(orderDiagCode).append('\'');
        sb.append(", orderDiagText='").append(orderDiagText).append('\'');
        sb.append(", orderComment='").append(orderComment).append('\'');
        sb.append(", orderDepartmentName='").append(orderDepartmentName).append('\'');
        sb.append(", orderDepartmentMisCode='").append(orderDepartmentMisCode).append('\'');
        sb.append(", orderDoctorFamily='").append(orderDoctorFamily).append('\'');
        sb.append(", orderDoctorName='").append(orderDoctorName).append('\'');
        sb.append(", orderDoctorPatronum='").append(orderDoctorPatronum).append('\'');
        sb.append(", orderDoctorMisId=").append(orderDoctorMisId);
        sb.append('}');
        return sb.toString();
    }

    public Staff getOrderDoctorMis() {
        return orderDoctorMis;
    }

    public void setOrderDoctorMis(Staff orderDoctorMis) {
        this.orderDoctorMis = orderDoctorMis;
    }
}
